clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -299,6 +363,34 @@ else if (expireMs <= now) {
         log.debug("Updated cluster metadata version {} to {}", this.version, this.cluster);
     }
 
+    /**
+     * Convert a {@link org.apache.kafka.common.requests.MetadataResponse.PartitionMetadata} into a
+     * {@link PartitionInfo}. If the epoch in the partition metadata is not newer than the last one we saw, use the
+     * previous partition info instance.
+     */
+    private PartitionInfo getPartitionInfoFromMeta(String topic, MetadataResponse.PartitionMetadata partitionMetadata) {
+        TopicPartition tp = new TopicPartition(topic, partitionMetadata.partition());
+        if (partitionMetadata.leaderEpoch().isPresent()) {
+            int newEpoch = partitionMetadata.leaderEpoch().get();
+            if (updateLastSeenEpoch(tp, newEpoch, oldEpoch -> newEpoch >= oldEpoch) == newEpoch) {
+                // If the received leader epoch is at least the same as the previous one, use the new partition info
+                return MetadataResponse.partitionMetaToInfo(topic, partitionMetadata);
+            } else {
+                // Otherwise ignore the new metadata and use the previously cached info
+                PartitionInfo previousInfo = cluster.partition(tp);
+                if (previousInfo != null) {
+                    return previousInfo;
+                } else {
+                    log.warn("Got an older epoch in partition metadata response for {}, but could not find previous partition " +
+                             "info to use so defaulting to the possible stale data from latest response", tp);
+                    return MetadataResponse.partitionMetaToInfo(topic, partitionMetadata);
```

Comment:
Hmm.. So in this case, the client was aware of a larger epoch, but didn't have the corresponding partition metadata. This could happen if the epoch was observed after reading a committed offset. Since we know the metadata we received from the response is stale, I wonder if it would be better to leave the cached metadata empty so that we will retry.

Reply:
Yea, could do that. Will refactor this a bit to avoid any nulls